### PR TITLE
feat: push down window last

### PR DESCRIPTION
**Pull request by jlapacik**

### CODE REVIEW

File: storage/reads/array_cursor.gen.go.tmpl [hunk 1]
```diff
@@ -242,16 +256,84 @@ func (c *{{.name}}LimitArrayCursor) Next() {{$arrayType}} {
...
 	every int64
 	final int64
 	res {{$arrayType}}
 	tmp {{$arrayType}}
 }
 
-func new{{.Name}}WindowLimitArrayCursor(cur cursors.{{.Name}}ArrayCursor, every int64) *{{.name}}WindowLimitArrayCursor {
-	return &{{.name}}WindowLimitArrayCursor{
+func new{{.Name}}WindowFirstArrayCursor(cur cursors.{{.Name}}ArrayCursor, every int64) *{{.name}}WindowFirstArrayCursor {
+	return &{{.name}}WindowFirstArrayCursor{
```
Comment (fchikwekwe) [inline storage/reads/array_cursor.gen.go.tmpl:336]: What did limit do?

Comment (jlapacik) [reply storage/reads/array_cursor.gen.go.tmpl:336]: Previously there was only a single "limit" array cursor for handling both last and first operations. But windowing needs different cursor implementations for first and last.

Comment (fchikwekwe) [reply storage/reads/array_cursor.gen.go.tmpl:336]: Okay, yeah that makes more sense now.

File: storage/reads/array_cursor.gen.go [hunk 2]
```diff
@@ -298,11 +324,79 @@ func newFloatWindowLimitArrayCursor(cur cursors.FloatArrayCursor, every int64) *
...
+		return c.res
+	}
+
+	for i, t := range a.Timestamps {
+		if t >= c.final {
+			cur++
+		}
+
+		if cur == MaxPointsPerBlock {
+			c.tmp.Timestamps = a.Timestamps[i+1:]
```
Comment (yzhang1991) [inline storage/reads/array_cursor.gen.go:360]: Imagine the following case where `every = 2m` and `MaxPointsPerBlock = 1`:
```
0, "value0"
1m, "value1"
2m, "value2"
```
The first window was `[0, 2m)`, we got the first output `1m, "value1"`, `c.final = 2m`, `cur = 0`
Then we encountered `2m, "value2"`. Because `t >= c.final`, we should increment `cur`  and that reached `MaxPointsPerBlock = 1`, then we assigned `tmp` to `data[i+1:]`. In this case `data[i]` is already the last record, so `data[i+1:]` is actually an empty slice.

The next time when we call `Next()`, `a.tmp.Len() == 0`, `c.FloatArrayCursor.Next()` also returns a 0-length array, so in line 349, you return `c.res.Timestamps[:cur+1]` and `c.res.Values[:cur+1]` which is basically `c.res.Timestamps[:0]` and `c.res.Values[:0]`, they are both empty.

Now, where is my `2m, "value2"`?

File: storage/reads/aggregate_resultset.go [hunk 3]
```diff
@@ -33,7 +33,13 @@ func NewWindowAggregateResultSet(ctx context.Context, req *datatypes.ReadWindowA
...
 
 	ascending := true
-	if len(req.Aggregate) > 0 && req.Aggregate[0].Type == datatypes.AggregateTypeLast {
+
+	// The following is an optimization where in the case of a single window,
+	// the selector `last` is implemented as a descending array cursor followed
+	// by a limit array cursor that selects only the first point, i.e the point
+	// with the largest timestamp, from the descending array cursor.
+	//
+	if req.Aggregate[0].Type == datatypes.AggregateTypeLast && (req.WindowEvery == 0 || req.WindowEvery == math.MaxInt64) {
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/aggregate_resultset.go:42]: Are you implying it should return 0 points, not 1 point?

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: Yes, because window end is exclusive.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: If my data is:
```
1, "value1"
math.MaxInt64, "value2"
```
`... |> window(every: inf) |> last()` should return `1, "value1"` not `math.MaxInt64, "value2"`

[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/aggregate_resultset.go:42]: Ah yes you are correct, this should only be checking for `req.WindowEvery == 0`.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: I think for the vast majority of the `window(every: inf)` cases, using `limit` is still more efficient? I just want to point out this one specific edge case that will not work.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: It will cause a problem only if you have `math.MaxInt64` timestamp row in your data but it is rare.

[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: To be honest, I don't really want to give up the reverse limit 1 optimization for `last` just because of 1 particular rare edge case, not sure what to suggest here.

[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/aggregate_resultset.go:42]: I originally used this as a reference:

https://github.com/influxdata/influxdb/blob/master/storage/reads/aggregate_resultset.go#L57-L68

But as you pointed out, this will not always be correct. That will have to be corrected as well. We also clearly need test coverage for this edge case.

File: storage/reads/array_cursor.gen.go [hunk 4]
```diff
@@ -44,26 +44,52 @@ func newLimitArrayCursor(cur cursors.Cursor) cursors.Cursor {
...
+}
+
+func newWindowLastArrayCursor(cur cursors.Cursor, every int64) cursors.Cursor {
+	if every == 0 {
+		return newLimitArrayCursor(cur)
+	}
+	switch cur := cur.(type) {
+
+	case cursors.FloatArrayCursor:
+		return newFloatWindowLastArrayCursor(cur, every)
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/array_cursor.gen.go:80]: This check is not necessary assuming `newWindowLastArrayCursor` is always called in the correct context. I could have removed the check and replaced it with a doc comment specifying that `every = 0` should never be passed in, but I decided it was just better to leave the check.

File: storage/reads/array_cursor.gen.go [hunk 5]
```diff
@@ -280,16 +306,16 @@ func (c *floatLimitArrayCursor) Next() *cursors.FloatArray {
...
 	tmp   *cursors.FloatArray
 }
 
-func newFloatWindowLimitArrayCursor(cur cursors.FloatArrayCursor, every int64) *floatWindowLimitArrayCursor {
-	return &floatWindowLimitArrayCursor{
+func newFloatWindowLastArrayCursor(cur cursors.FloatArrayCursor, every int64) *floatWindowLastArrayCursor {
+	return &floatWindowLastArrayCursor{
 		FloatArrayCursor: cur,
 		every:            every,
 		final:            math.MinInt64,
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/array_cursor.gen.go:321]: Perhaps a better name like `windowEnd` would be better instead?

File: storage/reads/array_cursor.gen.go [hunk 6]
```diff
@@ -298,11 +324,79 @@ func newFloatWindowLimitArrayCursor(cur cursors.FloatArrayCursor, every int64) *
...
 
-func (c *floatWindowLimitArrayCursor) Stats() cursors.CursorStats {
+func (c *floatWindowLastArrayCursor) Stats() cursors.CursorStats {
 	return c.FloatArrayCursor.Stats()
 }
 
-func (c *floatWindowLimitArrayCursor) Next() *cursors.FloatArray {
+func (c *floatWindowLastArrayCursor) Next() *cursors.FloatArray {
+	every := c.every
+	if every == 0 {
```
[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/array_cursor.gen.go:333]: No it is not possible for `every == 0` here if `newWindowLastArrayCursor()` is used. However if a WindowLastArrayCursor is instantiated in any other way, then a zero value for `every` could result in a panic inside this method. Therefore this was added purely as a precaution.

File: storage/reads/aggregate_resultset.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yzhang1991) [reply storage/reads/aggregate_resultset.go:42]: It turns out I forgot about the `range()` thing... I actually posted this https://github.com/influxdata/influxdb/pull/18414/files#r437653943 a while ago. So you can ignore my comments above.

[the remark this answers is not in the document]
Comment (jlapacik) [reply storage/reads/aggregate_resultset.go:42]: Ah yes, it turns out this edge case is actually invalid because influxdb's `range` requirement will always filter out any points at time `math.MaxInt64`.

File: storage/reads/array_cursor.gen.go (hunk 2 above)
Comment (jlapacik) [reply storage/reads/array_cursor.gen.go:360]: Typo.
```suggestion
			c.tmp.Timestamps = a.Timestamps[i:]
```

Comment (yzhang1991) [reply storage/reads/array_cursor.gen.go:360]: Some test coverage for it will be desirable

Comment (jlapacik) [reply storage/reads/array_cursor.gen.go:360]: Absolutely. Changes incoming...
